overlord/devicestate/handlers_install.go

```diff
@@ -618,3 +634,72 @@ func (m *DeviceManager) doRestartSystemToRunMode(t *state.Task, _ *tomb.Tomb) er
 
 	return nil
 }
+
+var seedOpen = seed.Open
+
+var maybeApplyPreseededData = func(ubuntuSeedDir, sysLabel, writableDir string) (preseeded bool, err error) {
+	preseedArtifact := filepath.Join(ubuntuSeedDir, "systems", sysLabel, "preseed.tgz")
+	if !osutil.FileExists(preseedArtifact) {
+		return false, nil
+	}
+
+	// TODO: verify artifact signature?
+
+	logger.Noticef("apply preseed data: %q, %q", writableDir, preseedArtifact)
+	cmd := exec.Command("tar", "--extract", "--preserve-permissions", "--preserve-order", "--gunzip", "--directory", writableDir, "-f", preseedArtifact)
+	if err := cmd.Run(); err != nil {
+		return false, err
+	}
+
+	logger.Noticef("copying snaps")
+
+	deviceSeed, err := seedOpen(dirs.SnapSeedDir, sysLabel)
+	if err != nil {
+		return false, err
+	}
+	tm := timings.New(nil)
+
+	// XXX: this is expensive, but we need snap revisions. Any way around this?
```

Comment:
If you cannot find better solutions, a suggestion I can give is to perform this in parallel with the extraction of the preseed archive. You could, for example, change `cmd.Run()` above with `cmd.Start()`, and add a `cmd.Wait()` after `LoadMeta()` below.

(this is assuming that `LoadAssertions` and `LoadMeta` don't need the preseed data, of course)

Reply:
I think we will want to avoid this altogether, e.g. once we agree on how to sign preseed.tgz artifact, we can store extra metadata inside and have the info ready there.

Reply:
That sounds like the smartest option indeed!

Reply:
This will change in a followup PR.